async/source_test.go

```diff
@@ -66,11 +68,65 @@ func TestConsumeMessagesSuccessfully(t *testing.T) {
 	}
 }
 
+func TestConsumeMessagesSuccessfullyConcurrently(t *testing.T) {
+	receivedAcks := make(chan substrate.Message)
+
+	var wg sync.WaitGroup
+
+	source := messageSourceAdapter{
+		source: &messageSourceMock{
+			consumerMessagesMock: func(ctx context.Context, messages chan<- substrate.Message, acks <-chan substrate.Message) error {
+				for i := 0; i < 100; i++ {
+					wg.Add(1)
+					messages <- message.FromString("payload")
+				}
+
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case ack := <-acks:
+						receivedAcks <- ack
+					}
+				}
+			},
+		},
+		consumers: 5,
+		msgBuffer: 100,
+		ackBuffer: 100,
+	}
+
+	sourceContext, sourceCancel := context.WithTimeout(context.Background(), time.Second)
+	defer sourceCancel()
+
+	errs := make(chan error)
+
+	go func() {
+		defer close(errs)
+		errs <- source.ConsumeMessages(sourceContext, func(_ context.Context, _ substrate.Message, ack AckFunc) error {
+			wg.Done()
```

Comment:
shouldn't we be adding to waitgroup for this goroutine?

Reply:
I mean we are adding waitgroup in https://github.com/uw-labs/substrate-tools/pull/6/files#diff-90006bcabbe2b6ae8a5c9a3d4157a6dbR80 but calling wg.Done() in different one, which is confusing me a bit.

Reply:
example usage https://gobyexample.com/waitgroups does this for the same goroutine

Reply:
even better in the example usage go docs https://golang.org/pkg/sync/#WaitGroup

Reply:
I've moved the `wg.Add()` call into the main routine for better readability.